### PR TITLE
Test

### PR DESCRIPTION
**Pull request by Nitr4x**
Adding unit test

### CODE REVIEW

**Comment by Nitr4x** [comment]
Please, for any contract related feedback, see https://github.com/Nitr4x/remittance/pull/2.

File: test/Owned.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        return instance.getOwner()
+            .then(owner => {
+                assert.strictEqual(owner, accounts[0]);
+            });
+    });
+
+    // Testing changeOwner function.
+    it('Owner should be ' + accounts[1], () => {
+        return instance.changeOwner(accounts[1], {from: accounts[0]})
+            .then(() => {
```
Comment (xavierlepretre) [inline test/Owned.js:25]: Testing Owned is not really the point of this repo here, but you could check the receipt too.

File: test/Remittance.js [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+    describe('======= newTransaction unit testing =======', () => {
+        it('Should create a new transaction with no extra fees', async () => {
+            const txObj = await instance.newTransaction(exchange, password, {from: owner, value: toWei("0.1", "ether")});
+            assert.strictEqual(txObj.logs.length, 1);
+            assert.strictEqual(txObj.logs[0].event, "LogNewTransaction");
+            assert.strictEqual(txObj.logs[0].args[0], owner);
+            assert.strictEqual(txObj.logs[0].args[1], exchange);
+            assert.strictEqual(fromWei(txObj.logs[0].args[2], "ether"), "0.1");
+
```
Comment (xavierlepretre) [inline test/Remittance.js]: Also test that storage is updated as expected.

File: test/Remittance.js [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+                instance.withdraw(password, {from: stranger})
+            );
+        });
+
+        it('Should withdraw the exchange balance', async () => {
+            let initialBalance = new BN(await web3.eth.getBalance(exchange));
+
+            const txObj = await instance.withdraw(password, {
+                from: exchange,
+                gasPrice: 50
```
Comment (xavierlepretre) [inline test/Remittance.js]: Have we seen how to do it without imposing the gas price?

**Review by xavierlepretre — COMMENTED** [review]
Beside the "public data" mistake in the contract code, it looks like you have the hang of things.

File: test/Remittance.js (hunk 2 above)
Comment (Nitr4x) [reply test/Remittance.js]: How can I check the storage considering that the map is private?

File: test/Remittance.js (hunk 3 above)
Comment (Nitr4x) [reply test/Remittance.js]: I must admit that I don't remember this subtility

File: test/Remittance.js (hunk 2 above)
Comment (xavierlepretre) [reply test/Remittance.js]: If you want to make it easy, make it `public`.
If you want to make it difficult, look at how the location is calculated with `sha3`, and then use `web3.eth.getStorageAt`.

File: test/Remittance.js (hunk 3 above)
Comment (xavierlepretre) [reply test/Remittance.js]: You let the system choose the gas price for you, then, at calculation time, you look at what was picked for you.

File: test/Remittance.js [hunk 4]
```diff
@@ -122,6 +122,20 @@ contract('Remittance', (accounts) => {
             assert.strictEqual(txObj.logs[1].args[2].toString(), new BN(toWei("0.1", "ether")).minus(2000).toString());
         });
 
+        it('Should be able to cancel a transaction after one 1 day time', async () => {
+            hashedOTP = await instance.hashOTP(owner, password);
+            await instance.newTransaction(owner, hashedOTP, 1, {from: owner, value: toWei("0.1", "ether")});
```
Comment (xavierlepretre) [inline test/Remittance.js]: That's a 1 second delay.

Comment (Nitr4x) [reply test/Remittance.js]: Sorry for this stupid error...

File: test/Remittance.js [hunk 5]
```diff
@@ -0,0 +1,339 @@
+'use strict';
+
+const Remittance = artifacts.require('Remittance');
+const truffleAssert = require('truffle-assertions');
+
+const {toWei, fromWei } = web3.utils;
```
Comment (xavierlepretre) [inline test/Remittance.js]: `const { toWei, fromWei, BN, utf8ToHex } = web3.utils;`

File: test/Remittance.js [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+            );          
+        });
+
+        it('Should fail if the tx value is equal to 0', async () => {
+            await truffleAssert.reverts(
+                instance.newTransaction(exchange, password, fiveDays, {from: owner})
+            );          
+        });
+
+        it('Should fail if the hashedOTP is already used', async () => {
```
Comment (xavierlepretre) [inline test/Remittance.js]: Is currently in use.

File: test/Remittance.js [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+        beforeEach('Creating new transaction', async () => {
+            hashedOTP = await instance.hashOTP(exchange, password);
+
+            const txObj = await instance.newTransaction(exchange, hashedOTP, fiveDays, {from: owner, value: toWei("0.1", "ether")});
+
+            assert.strictEqual(txObj.logs.length, 2);
+            assert.strictEqual(txObj.logs[0].event, "LogTakeFees");
+            assert.strictEqual(txObj.logs[0].args[0], owner);
+            assert.strictEqual(txObj.logs[0].args[1].toString(), "2000");
+            assert.strictEqual(txObj.logs[1].event, "LogNewTransaction");
```
Comment (xavierlepretre) [inline test/Remittance.js]: All this was already tested elsewhere. Better leave the `beforeEach` do straight actions.

File: test/Remittance.js [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+            assert.strictEqual(txObj.logs[1].args[0], owner);
+            assert.strictEqual(txObj.logs[1].args[1], exchange);
+            assert.strictEqual(txObj.logs[1].args[2].toString(), new BN(toWei("0.1", "ether")).minus(2000).toString());
+        });
+
+        it('Should be able to cancel a transaction after one 1 day time', async () => {
+            hashedOTP = await instance.hashOTP(owner, password);
+            await instance.newTransaction(owner, hashedOTP, 1 * 24 * 60 * 60, {from: owner, value: toWei("0.1", "ether")});
```
Comment (xavierlepretre) [inline test/Remittance.js]: Why not use the one of `beforeEach`?

File: test/Remittance.js [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+
+            await increaseTime(1 * 24 * 60 * 60);
+    
+            let txObj = await instance.cancelTransaction(owner, hashedOTP, {from: owner});
+            assert.strictEqual(txObj.logs.length, 1);
+            assert.strictEqual(txObj.logs[0].event, "LogCancelTransaction");
+            assert.strictEqual(txObj.logs[0].args[0], owner);
+            assert.strictEqual(txObj.logs[0].args[1], owner);
+            assert.strictEqual(txObj.logs[0].args[2].toString(), new BN(toWei("0.1", "ether")).minus(2000).toString());
```
Comment (xavierlepretre) [inline test/Remittance.js]: What about `toWei(new BN("0.1"), "ether")`

File: test/Remittance.js [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+
+        beforeEach('Creating new transaction', async () => {
+            hashedOTP = await instance.hashOTP(exchange, password);
+
+            const txObj = await instance.newTransaction(exchange, hashedOTP, fiveDays, {from: owner, value: toWei("0.1", "ether")});
+
+            assert.strictEqual(txObj.logs.length, 2);
+            assert.strictEqual(txObj.logs[0].event, "LogTakeFees");
+            assert.strictEqual(txObj.logs[0].args[0], owner);
+            assert.strictEqual(txObj.logs[0].args[1].toString(), "2000");
```
Comment (xavierlepretre) [inline test/Remittance.js]: Same here about these tests in `beforeEach`.

File: test/Remittance.js [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+        });
+
+        it('Should allow various person to withdraw fees', async () => {
+            let txObj = await instance.changeOwner(stranger, {from: owner});
+            
+            assert.strictEqual(txObj.logs.length, 1);
+            assert.strictEqual(txObj.logs[0].event, "LogOwnerChanged");
+            assert.strictEqual(txObj.logs[0].args[0], stranger);
+
+            await instance.newTransaction(exchange, await instance.hashOTP(exchange, web3.utils.utf8ToHex("haha")), fiveDays, {from: owner, value: toWei("0.1", "ether")});
```
Comment (xavierlepretre) [inline test/Remittance.js]: Make it 2 lines? :)

File: test/Remittance.js (hunk 5 above)
Comment (Nitr4x) [reply test/Remittance.js]: Not agreeing for the BN utility. Indeed I have some bug when calling BN functions. That's why I used the package big-number

File: test/Remittance.js (hunk 7 above)
Comment (Nitr4x) [reply test/Remittance.js]: OK I was not confident about this point. Thanks to cleared this out :)

File: test/Remittance.js (hunk 8 above)
Comment (Nitr4x) [reply test/Remittance.js]: Cause the hashedOTP of beforeEach is currently in use and was defined with a higher delay

File: test/Remittance.js (hunk 9 above)
Comment (Nitr4x) [reply test/Remittance.js]: Cause I need to perform operation. Moreover, your way of doing it is triggering an error. Waht is wrong by using this method? Why should I changed it?

Comment (xavierlepretre) [reply test/Remittance.js]: I was hoping that my syntax would be clearer. It ends in `)` instead of `))`. :)

**Comment by Nitr4x** [comment]
Too much comments. Creating new pull request.

**Pull request closed without merging**